Kuzmenko/tag cloud component

kusem opened the pull request:
Updates:
- tag cloud component
- "searches" table
- fake data seeder

kusem committed:
Add Tag Cloud Component.
Add "searches" table.
Add fake data seeder (Fixture) for "searches"

kusem committed:
Add new lines at EOF.

TheVitik commented on src/Controller/TagCloudController.php:
```diff
@@ -0,0 +1,47 @@
+            'TagCloud.html.twig',['topTags'=>(array) $topTags]
+        );
+    }
+}
```
Add newline at end of file

TheVitik commented on templates/TagCloud.html.twig:
```diff
@@ -0,0 +1,21 @@
+            </div>
+        </div>
+    </div>
+{% endblock %}
```
Add newline at end of file

TheVitik commented on templates/TagCloud.html.twig:
```diff
@@ -0,0 +1,21 @@
+{% extends 'base.html.twig' %}
+{% block content %}
```
Make a separate template for tag cloud rendering

kusem replied:
Not needed, Tag Cloud shouldn't be placed on every page according to the initial task.

TheVitik commented on src/Controller/TagCloudController.php:
```diff
@@ -0,0 +1,47 @@
+    public function getTopTags(int $qty = 50, SearchesRepository $searches){
+        if ($qty<0) throw new InvalidArgumentException('Bad quantity argument while calling getTopTags function. Argument should be higher than 0.');
+
+        $query = $searches->createQueryBuilder()
```
In my opinion, a database query should be in a separate function in SearchesRepository.

kusem committed:
Add caching.
Add index searches.word
Fix .env privacy :)

kusem committed:
Move cache to Repository
Change Cache settings (global lifetime of topTag cache)
Move component templates to "blocks" folder
Delete HomeController as not needed (because Viktor finish normal page)

TheVitik reviewed (approved):
Good job

taras-koval commented on src/Controller/TagCloudController.php:
```diff
@@ -0,0 +1,29 @@
+     * @return Response
+     * @throws InvalidArgumentException
+     */
+    #[Route('/TagCloud', name: 'TagCloud')]
```
route with capital letters looks strange

kusem replied:
Fixed, thanks.

taras-koval commented on src/Controller/TagCloudController.php:
```diff
@@ -0,0 +1,29 @@
+    {
+        $topTags = $repository->getTopTags();
+
+        return $this->render('TagCloud.php.twig', [
```
use snake case for twig filenames (https://symfony.com/doc/current/templates.html#template-naming)

taras-koval commented on src/Entity/Searches.php:
```diff
@@ -0,0 +1,50 @@
+    #[ORM\Id]
+    #[ORM\GeneratedValue]
+    #[ORM\Column]
+    private ?int $id = null;
```
in my opinion, the fields in this table should be "not null"

kusem replied:
> in my opinion, the fields in this table should be "not null"

as for me, these validations should be made on the "insert side" (before insert queries): "id" will not be null in any case (its AI), "word" should be checked before insert, "cnt" also should be set before inserting. I do not see any reasons to add one more "layer" of protections for this table but can add it if someone else does not agree with me.

taras-koval commented on templates/blocks/_TagCloud.html.twig:
```diff
@@ -0,0 +1,21 @@
+{#  {% extends 'base.html.twig' %} #}
+{% block content %}
+    <style>
```
it is better to put styles in a separate file

kusem replied:
> it is better to put styles in a separate file

Yeah, forgot to move them after testing. Thanks, will fix it in the nearest commit.

taras-koval commented on templates/TagCloud.php.twig:
```diff
@@ -0,0 +1,6 @@
+
+{% block content %}
+    <h3>The most popular searches</h3>
+    {{ include('blocks/_TagCloud.html.twig') }}
```
use snake case for twig filenames (https://symfony.com/doc/current/templates.html#template-naming)

kusem committed:
Changes in route naming for "tag cloud"
Move style from 'twig' to separate css file

taras-koval commented on src/Controller/TagCloudController.php:
```diff
@@ -17,12 +17,12 @@ class TagCloudController extends AbstractController
      * @throws InvalidArgumentException
      */
-    #[Route('/TagCloud', name: 'TagCloud')]
+    #[Route('/tag-cloud', name: 'tag-cloud')]
```
use snake case for route name (examples https://symfony.com/doc/current/routing.html#creating-routes)
`#[Route('/tag-cloud', name: 'tag_cloud')]`

kusem replied:
> use snake case for route name (examples https://symfony.com/doc/current/routing.html#creating-routes) `#[Route('/tag-cloud', name: 'tag_cloud')]`

Do not see any obliges to use snake case or kebab case - they use both of them in their examples.

taras-koval replied:
for route path use "/tag-cloud", for route name use "tag_cloud"
![image](https://user-images.githubusercontent.com/26673933/197065275-70fd48dc-b4fb-4ef6-ac4c-09185009100f.png)

kusem replied:
Changed.

kusem committed:
s$snake_case for Taras

kusem committed:
Changed route pass to kebab style, route name to snake case.

yderevianko-cartera commented on src/Controller/HomeController.php:
```diff
@@ -1,24 +0,0 @@
-<?php
-
-namespace App\Controller;
```
Why are you deleting HomeController ?

kusem replied:
> Why are you deleting HomeController ?

Just because I do not need it anymore, and to avoid any route troubles which might appear after Viktor finished his index page task.

yderevianko-cartera replied:
if you merge it will delete it from main branch. 
First use 
git pull --rebase origin master
and
git push -f

yderevianko-cartera commented on src/Repository/SearchesRepository.php:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    public function save(Searches $entity, bool $flush = false): void
+    {
```
Repositiry should only have queries that return something.
I do not thing that you need methods like save and remove since you can use EntittyManager in Controller to save something.

kusem replied:
These methods are generated by Symfony automatically.

Ok, I will delete them.

yderevianko-cartera commented on src/Repository/SearchesRepository.php:
```diff
@@ -0,0 +1,90 @@
+//     * @return Searches[] Returns an array of Searches objects
+//     */
+//    public function findByExampleField($value): array
+//    {
```
please remove comments

kusem committed:
Replace "handmade" style for buttons margin using bootstrap style m-1.
Return HomeController.php into family.
Replace /search/{word} with /search?q={word} in tag cloud href

kusem committed:
Move caching into TagCloudController.php
